### PR TITLE
Tetromino handling algorithm. Close #18

### LINKED ISSUE #18: Tetromino handling algorithm
After all helper functions in place we can implement class-based operations on tetromino, namely rotation and translation.
It should remember current rotation phase and check boundary conditions.
Create a class for each of tetromino types (7 classes).

You should update/access inner properties of class via constructor and methods:
- phase
- coordinates
- tetromino type (this should not change after initialization)

## Constructor
Each class should have a constructor, that will create tetromino of this type. 

Constructor should have 3 parameters:
- tetromino type (single letter)
- initial rotationPhase (string, like "horizontal")
- initial coordinates (array of (x,y)-objects) (in tests use buildTetromino)

## Rotate tetromino
There should be a method to rotate tetromino on a field

Parameters:
- field (array of arrays of booleans). If element is true - this coordinate already has a block on it. Size of outer array - is height of field (y-coordinate). Size of inner array - is width of field (x-coordinate). In tests, please construct parameter via buildField.

Function returns:
- coordinates property of class

Algorithm:
- Initiate rotation (rotateTetromino - all parameters should come from inside the class itself: typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates)
- check if valid position - not out-of-bounds or intersecting places blocks on field (checkThatTheFieldIsFree - move to top-level function in tetris.js, add tests for it in separate test file)
- If check result is valid - update rotation phase (incrementPhase) and coordinates
- If check result is invalid - leave tetromino coordinates and phase as they were before rotation was attempted

Unit tests
- for each type of tetromino, create it and cycle through all rotation phases and check that it returns to the original position. (7 tests)
Cover with tests following equivalence classes:
- different tetromino types
- rotation starts in different phases
- intersects with blocks on field after rotation
- goes out of bounds of field
  * too low
  * too high 
  * too to the left
  * too to the right

## Move tetromino
This function will translate tetromino position.

Parameters:
- field (same as above)
- shift - (x,y)-object. Function should do nothing if invalid move is attempted. Valid moves (shifts):
- 1 to the left (-1,0)
- 1 to the right (1,0)
- 1 down (0,1)

Function returns:
- coordinates property of class

Algorithm:
- Initiate movement - (shiftCoordinates)
- check if valid position - not out-of-bounds or intersecting places blocks on field (checkThatTheFieldIsFree)
- If check result is valid - update coordinates
- If check result is invalid - do not update coordinates

Unit tests:
One unit test for each type of tetromino for valid movement (21 tests)
- left 
- right
- down

One unit test for each type of tetromino for invalid movement:
- goes out of bounds of field
  * too low
  * too to the left
  * too to the right

### PR DESCRIPTION
**Pull request by PashaBeshchuk**

### CODE REVIEW

File: test/test-rotatesTetrominos.js [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        tetris.incrementPhase(){
+            
+        }
+        tetris.finelRotationTetromino(field){
+
+        }
+            
+    }
+    it("L-tetromino rotation", function(){
+        let typeTetramoni = "L";
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Tetramoni?

File: test/test-rotatesTetrominos.js [hunk 2]
```diff
@@ -0,0 +1,32 @@
+var assert = require('chai').assert;
+var tetris = require('../tetris');
+describe("rotate L-tetromino", function (){
+    class Tetramino = {
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: This should not be part of test

File: test/test-rotatesTetrominos.js [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+        constructor(tetramino, phase, coordinates){
+            this.typeTetramoni = tetramino
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Typo.

File: test/test-rotatesTetrominos.js [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+            this.initialRotationPhase = phase;
+            this.initialCoordinates = coordinates;
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Parameters are "initial", not properties. Also for other properties

File: test/test-rotatesTetrominos.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+        }
+        tetris.rotateTetromino(){
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: This is class method. This is not components of tetris.
Also other methods. 
See how methods are defined in classes

File: test/test-rotatesTetrominos.js [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    it("L-tetromino rotation", function(){
+        let typeTetramoni = "L";
+        let coordinatesTetromino = tetris.getCoordinates(`
+            -----
+            ----X
+            --XXX
+            -----
+        `);
+        let phase = "rihgt";
+        let L-tetromino = class Tetramino(typeTetramoni, phase, coordinatesTetromino);
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: This is not how objects are created

Comment (zbstof) [reply test/test-rotatesTetrominos.js]: L-tetramino is not valid variable name. Dashes are not allowed
Also should start with lowercase

File: test/test-rotatesTetrominos.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    }
+    it("L-tetromino rotation", function(){
+        let typeTetramoni = "L";
+        let coordinatesTetromino = tetris.getCoordinates(`
+            -----
+            ----X
+            --XXX
+            -----
+        `);
+        let phase = "rihgt";
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Typo

File: test/test-rotatesTetrominos.js (hunk 6 above)
Comment (zbstof) [reply test/test-rotatesTetrominos.js]: Tertamino should be base type, L tetromino should be created from class L

File: test/test-rotatesTetrominos.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+            this.initialRotationPhase = phase;
+            this.initialCoordinates = coordinates;
+        }
+        tetris.rotateTetromino(){
+					
+        }
+        tetris.incrementPhase(){
+            
+        }
+        tetris.finelRotationTetromino(field){
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: This method is not in the requirements

File: test/test-rotatesTetrominos.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+    class Tetramino = {
+        constructor(tetramino, phase, coordinates){
+            this.typeTetramoni = tetramino
+            this.initialRotationPhase = phase;
+            this.initialCoordinates = coordinates;
+        }
+        tetris.rotateTetromino(){
+					
+        }
+        tetris.incrementPhase(){
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: This method is not in the requirements

File: test/test-rotatesTetrominos.js [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+        let typeTetramoni = "L";
+        let coordinatesTetromino = tetris.getCoordinates(`
+            -----
+            ----X
+            --XXX
+            -----
+        `);
+        let phase = "rihgt";
+        let L-tetromino = class Tetramino(typeTetramoni, phase, coordinatesTetromino);
+    })
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Where is the assert statement? This test doesn't check the output

File: test/test-rotatesTetrominos.js (hunk 2 above)
Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: How I will be test?

File: test/test-rotatesTetrominos.js (hunk 4 above)
Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: i do not understand, what is wrong here? :(

Comment (zbstof) [reply test/test-rotatesTetrominos.js]: You have field named initialCoordinates , should be "coordinates" .
You have constructor parameter named "coordinates", should be initialCoordinates
And for other fields/parameters

File: test/test-rotatesTetrominos.js (hunk 2 above)
Comment (zbstof) [reply test/test-rotatesTetrominos.js]: How do you test methods? 
You put module name in front and a dot

File: test/test-rotatesTetrominos.js [hunk 11]
```diff
@@ -1,32 +1,32 @@
...
-        tetris.rotateTetromino(){
+        _rotateTetromino(){
 					
         }
-        tetris.incrementPhase(){
+        _incrementPhase(){
             
         }
-        tetris.finelRotationTetromino(field){
+        _finelRotationTetromino(field){
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Typo

File: test/test-rotatesTetrominos.js [hunk 12]
```diff
@@ -1,32 +1,32 @@
...
+        let typeTetromino = "L";
         let coordinatesTetromino = tetris.getCoordinates(`
             -----
             ----X
             --XXX
             -----
         `);
         let phase = "rihgt";
-        let L-tetromino = class Tetramino(typeTetramoni, phase, coordinatesTetromino);
+        //let LTetromino = class Tetramino(typeTetramoni, phase, coordinatesTetromino);
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Don't commit commented code

File: tetris.js [hunk 13]
```diff
@@ -203,6 +183,32 @@ module.exports = {
 		}
 
 	},
+
+	checkThatTheFieldIsFree: function (arrayOfCoordinates, localField, lengthField_X, lengthField_Y) {
```
Comment (zbstof) [inline tetris.js]: Why are you passing dimensions? You can get them from field parameter

File: tetris.js [hunk 14]
```diff
@@ -203,6 +183,32 @@ module.exports = {
...
+			}
+		}
+		return true;
+	},
+
+	allItemsOnTheField: function (arrayOfCoordinates, lengthField_X, lengthField_Y) {
+		for (let coordinate of arrayOfCoordinates) {
+			if (
+				(coordinate.x < 0 || coordinate.y < 0) ||
+				(coordinate.x > (lengthField_X - 1)) ||
```
Comment (zbstof) [inline tetris.js]: Use >= to avoid -1

File: tetris.js [hunk 15]
```diff
@@ -203,6 +183,32 @@ module.exports = {
...
+				(coordinate.y > (lengthField_Y - 1))
```
Comment (zbstof) [inline tetris.js]: And here

File: tetris.js [hunk 16]
```diff
@@ -203,6 +183,32 @@ module.exports = {
 		}
 
 	},
+
+	checkThatTheFieldIsFree: function (arrayOfCoordinates, localField, lengthField_X, lengthField_Y) {
+		if (this.allItemsOnTheField(arrayOfCoordinates, lengthField_X, lengthField_Y) === false) {
```
Comment (zbstof) [inline tetris.js]: Dont compare booleans to booleans
Use negation operator

File: tetris.js [hunk 17]
```diff
@@ -203,6 +183,32 @@ module.exports = {
...
+			return false
+		}
+		for (let coordinate of arrayOfCoordinates) {
+			if (localField[coordinate.y][coordinate.x] === true) {
```
Comment (zbstof) [inline tetris.js]: Just use boolean value directly
Comparisin is not needed

File: test/test-rotatesTetrominos.js [hunk 18]
```diff
@@ -1,32 +1,194 @@
 var assert = require('chai').assert;
 var tetris = require('../tetris');
-describe("rotate L-tetromino", function (){
+describe("Class tetramino", function (){
     class Tetromino {
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: again: move this outside of test

File: test/test-rotatesTetrominos.js [hunk 19]
```diff
@@ -1,32 +1,194 @@
...
-        constructor(tetramino, phase, coordinates){
-            this.typeTetramoni = tetramino
-            this.initialRotationPhase = phase;
-            this.initialCoordinates = coordinates;
+        constructor(typeTetramoni, initialRotationPhase, initialCoordinates){
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: typo in typeTetramoni

File: test/test-rotatesTetrominos.js [hunk 20]
```diff
@@ -1,32 +1,194 @@
...
+            this.typeTetramoni = typeTetramoni
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: typo

File: test/test-rotatesTetrominos.js [hunk 21]
```diff
@@ -1,32 +1,194 @@
...
+            let newCoordinates = this.rotatesArrayOfCoordinates(tetrominoCoordinates, pivot)
+            let result = this.shiftCoordinates(newCoordinates, shift)
+            return result.sort(this.orderCoordinates);
         }
-        _incrementPhase(){
-            
-        }
-        _finelRotationTetromino(field){
 
+        incrementPhase (typeOfTetromino, currentRotationPhaseOfTetromino) {
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: format

File: test/test-rotatesTetrominos.js [hunk 22]
```diff
@@ -1,32 +1,194 @@
...
+            if (typeOfTetromino === "I" || typeOfTetromino === "S" || typeOfTetromino === "Z") {
+                arrayPhases = ["vertical", "horizontal"];
+            }
+            if (typeOfTetromino === "L" || typeOfTetromino === "J" || typeOfTetromino === "T") {
+                arrayPhases = ["up", "left", "down", "right"];
+            }
+            if (typeOfTetromino === "O") {
+                arrayPhases = ["stable"];
+            }
+            let numberPhase = arrayPhases.indexOf(currentRotationPhaseOfTetromino)
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: phaseNumber

File: test/test-rotatesTetrominos.js [hunk 23]
```diff
@@ -1,32 +1,194 @@
...
+            if (numberPhase !== arrayPhases.length - 1) {
+                return arrayPhases[numberPhase + 1]
+            } else {
+                return arrayPhases[0]
+            }
         }
-            
+
+        determineShift (typeOfTetromino, rotationPhaseOfTetromino) {
+            if (typeOfTetromino === "O") {
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: what's all this? you already have this functionality in tetris.determineShift

File: test/test-rotatesTetrominos.js [hunk 24]
```diff
@@ -1,32 +1,194 @@
...
+            let result = this.shiftCoordinates(newCoordinates, shift)
+            return result.sort(this.orderCoordinates);
         }
-        _incrementPhase(){
-            
-        }
-        _finelRotationTetromino(field){
 
+        incrementPhase (typeOfTetromino, currentRotationPhaseOfTetromino) {
+            let arrayPhases = [];
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: you already have this functionality in tetris.incrementPhase

File: test/test-rotatesTetrominos.js [hunk 25]
```diff
@@ -1,32 +1,194 @@
...
-            this.initialCoordinates = coordinates;
+        constructor(typeTetramoni, initialRotationPhase, initialCoordinates){
+            this.typeTetramoni = typeTetramoni
+            this.initialRotationPhase = initialRotationPhase;
+            this.initialCoordinates = initialCoordinates;
         }
-        _rotateTetromino(){
-					
+        rotateTetromino(typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates) {
+            let shift = this.determineShift(typeOfTetromino, rotationPhaseOfTetromino)
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: you already have this functionality in tetris.rotateTetromino
Your algorithm is different

File: test/test-rotatesTetrominos.js (hunk 21 above)
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: this method is not in the requirements, remove

File: test/test-rotatesTetrominos.js [hunk 26]
```diff
@@ -1,32 +1,194 @@
...
+            let numberPhase = arrayPhases.indexOf(currentRotationPhaseOfTetromino)
+            if (numberPhase !== arrayPhases.length - 1) {
+                return arrayPhases[numberPhase + 1]
+            } else {
+                return arrayPhases[0]
+            }
         }
-            
+
+        determineShift (typeOfTetromino, rotationPhaseOfTetromino) {
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: this method is not in the requirements, remove

File: test/test-rotatesTetrominos.js [hunk 27]
```diff
@@ -1,32 +1,194 @@
...
+                }
+                if (rotationPhaseOfTetromino === "vertical") {
+                    return { x: -1, y: 1 };
+                }
+            }
+        } 
     }
     it("L-tetromino rotation", function(){
         let typeTetromino = "L";
+        let phase = "rihgt";
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: typo

File: test/test-rotatesTetrominos.js [hunk 28]
```diff
@@ -1,32 +1,194 @@
...
+    it("O-tetromino rotation", function(){
+        let typeTetromino = "O";
+        let phase = "stable";
+        let coordinatesTetromino = tetris.getCoordinates(`
+            -----
+            -XX--
+            -XX--
+            -----
+        `);
+        let TTetromino = new Tetromino(typeTetromino, phase, coordinatesTetromino)
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: variable names should start with lowercase letters. In all other tests as well

File: test/test-rotatesTetrominos.js (hunk 23 above)
Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: I do not understand, can I use methods from a tetris file?

File: test/test-rotatesTetrominos.js (hunk 27 above)
Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: Where is the typo?
All words is correct

File: test/test-rotatesTetrominos.js [hunk 29]
```diff
@@ -1,114 +1,42 @@
...
+    tetris.rotateTetromino(typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates) {
+        let shift = this.determineShift(typeOfTetromino, rotationPhaseOfTetromino)
+        let pivot = this.calculatorCoordinateOfTopLeft(tetrominoCoordinates);
+        let newCoordinates = this.rotatesArrayOfCoordinates(tetrominoCoordinates, pivot)
+        let result = this.shiftCoordinates(newCoordinates, shift)
+        return result.sort(this.orderCoordinates);
+    }
+
+    tetris.incrementPhase(typeOfTetromino, currentRotationPhaseOfTetromino) {
+        let arrayPhases = [];
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: why do you need lines 18-33? You already have them in tetris.incrementPhase

File: test/test-rotatesTetrominos.js [hunk 30]
```diff
@@ -1,114 +1,42 @@
 var assert = require('chai').assert;
 var tetris = require('../tetris');
-describe("Class tetramino", function (){
-    class Tetromino {
-        constructor(typeTetramoni, initialRotationPhase, initialCoordinates){
-            this.typeTetramoni = typeTetramoni
-            this.initialRotationPhase = initialRotationPhase;
-            this.initialCoordinates = initialCoordinates;
+class Tetromino {
```
[the remark this answers is not in the document]
Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: I must create new file and move class to another file?

[the remark this answers is not in the document]
Comment (zbstof) [reply test/test-rotatesTetrominos.js]: Move to tetris.js

File: tetris.js [hunk 31]
```diff
@@ -431,4 +431,15 @@ module.exports = {
...
\ No newline at end of file
+	class Tetromino {
+		constructor(typeOfTetromino, initialRotationPhase, initialCoordinates) {
+			this.typeOfTetromino = typeOfTetromino
+			this.initialRotationPhase = initialRotationPhase;
+			this.initialCoordinates = initialCoordinates;
+		}
+		tetris.rotateTetromino(typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates)
+	
+		tetris.incrementPhase(typeOfTetromino, currentRotationPhaseOfTetromino)
```
Comment (zbstof) [inline tetris.js]: this function is not in the acceptance criteria, please remove

File: tetris.js [hunk 32]
```diff
@@ -431,4 +431,15 @@ module.exports = {
...
 	}
-}
\ No newline at end of file
+	class Tetromino {
+		constructor(typeOfTetromino, initialRotationPhase, initialCoordinates) {
+			this.typeOfTetromino = typeOfTetromino
+			this.initialRotationPhase = initialRotationPhase;
+			this.initialCoordinates = initialCoordinates;
+		}
+		tetris.rotateTetromino(typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates)
```
Comment (zbstof) [inline tetris.js]: parameters typeOfTetromino, rotationPhaseOfTetromino, tetrominoCoordinates are not in the acceptance criteria, please remove them

File: test/test-rotatesTetrominos.js [hunk 33]
```diff
@@ -13,6 +13,7 @@ describe("Class tetramino", function () {
             -----
         `);
         let lTetromino = new Tetromino(typeOfTetromino, phase, coordinatesTetromino)
+        let rotateTetromino = finalTetraminoRotate(lTetromino.typeOfTetromino, lTetromino.typeOfTetromino, lTetromino.initialCoordinates)
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: finalTetraminoRotate - you don't have such method

Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: Shoud I improve the function rotateTetramino?

File: test/test-rotatesTetrominos.js [hunk 34]
```diff
@@ -0,0 +1,101 @@
...
+        let typeOfTetromino = "L";
+        let phase = "rihgt";
+        let coordinatesTetromino = tetris.getCoordinates(`
+            -----
+            ----X
+            --XXX
+            -----
+        `);
+        let lTetromino = new Tetromino(typeOfTetromino, phase, coordinatesTetromino)
+        let rotateTetrominoUp    = tetris.rotateTetromino(lTetromino.typeOfTetromino, lTetromino.typeOfTetromino, lTetromino.initialCoordinates)
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Align parameter list with requirements

File: test/test-rotatesTetrominos.js (hunk 7 above)
Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: Is the word phase incorect

File: test/test-rotatesTetrominos.js [hunk 35]
```diff
@@ -60,7 +95,18 @@ describe("Class tetramino", function () {
             --XX-
             -----
         `);
-        let zTetromino = new Tetromino(typeOfTetromino, phase, coordinatesTetromino)
+        let zTetromino = new tetris.Tetromino(typeOfTetromino, phase, coordinatesTetromino)
+        let rotateTetrominoVertical    = tetris.rotateTetromino(zTetromino.typeOfTetromino, zTetromino.typeOfTetromino, zTetromino.initialCoordinates)
+        let rotateTetrominoHorizontal  = tetris.rotateTetromino(zTetromino.typeOfTetromino, zTetromino.typeOfTetromino, zTetromino.initialCoordinates)
+        let rotateTetrominoVerticalSecond    = tetris.rotateTetromino(zTetromino.typeOfTetromino, zTetromino.typeOfTetromino, zTetromino.initialCoordinates)
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Just once is enough

File: test/test-rotatesTetrominos.js [hunk 36]
```diff
@@ -0,0 +1,180 @@
...
+        let phase = "stable";
+        let coordinatesTetromino = tetris.getCoordinates(`
+            -----
+            -XX--
+            -XX--
+            -----
+        `);
+        let oTetromino = new tetris.Tetromino(typeOfTetromino, phase, coordinatesTetromino)
+        let rotateTetrominoFirst  = tetris.rotateTetromino(iTetromino.typeOfTetromino, iTetromino.typeOfTetromino, iTetromino.initialCoordinates)
+        let rotateTetrominoSecond = tetris.rotateTetromino(iTetromino.typeOfTetromino, iTetromino.typeOfTetromino, iTetromino.initialCoordinates)
```
Comment (zbstof) [inline test/test-rotatesTetrominos.js]: Just once is enough

File: test/test-rotatesTetrominos.js [hunk 37]
```diff
@@ -177,4 +201,335 @@ describe("Class tetramino", function () {
...
+        let rotateTetromino = tetris.rotateTetromino(field)
+        let expected = tetris.getCoordinates(`
+            -----
+            -----
+            -XXXX
+            -----
+        `);
+        assert.deepEqual(rotateTetromino, expected)
+    })
+    it("O tetromino can not rotate, too down to the edge and field have element", function (){
```
[the remark this answers is not in the document]
Comment (PashaBeshchuk) [reply test/test-rotatesTetrominos.js]: let it be :)

[the remark this answers is not in the document]
Comment (zbstof) [reply test/test-rotatesTetrominos.js]: Ok

**Comment by zbstof** [comment]
Now do implementation

**Comment by zbstof** [comment]
Now implement shift function and tests in this PR

File: test/test-moveTetromino.js [hunk 38]
```diff
@@ -0,0 +1,92 @@
...
+            -----
+        `)
+        let field = tetris.buildField(`
+            -----
+            -----
+            -----
+            -----
+        `);
+        let lTetromino = new tetris.Tetromino(typeOfTetromino, phase, coordinatesTetromino);
+        let resultMoveTetromino = lTetromino.moveTetromino(coordinates, field);
```
Comment (zbstof) [inline test/test-moveTetromino.js]: This is not correct. 
Read requirements carefully

Comment (PashaBeshchuk) [reply test/test-moveTetromino.js]: Is this function have should accept three parameters(typeOfTetromino, coordinatesTetromino, field)?

Comment (zbstof) [reply test/test-moveTetromino.js]: Не вгадуй, будь ласка

Comment (PashaBeshchuk) [reply test/test-moveTetromino.js]: По описанию задачи, я пришел к выводу что должно быть два параметра, одни с координатами функции, другой с полем. Координаты - сдвигать, поле что бы проверить пусто оно или нет. 
Функция должна быть внутри класса Тетрамино, по моему я все так и сделал)

Comment (zbstof) [reply test/test-moveTetromino.js]: Incorrect. Read more carefully.

Comment (PashaBeshchuk) [reply test/test-moveTetromino.js]: Я верно понял, что тетромнино должно двигаться вниз, постоянно, как в игре?

Comment (zbstof) [reply test/test-moveTetromino.js]: Where in the issue does it say that?

File: tetris.js [hunk 39]
```diff
@@ -104,7 +76,7 @@ module.exports = {
 					}
 					newCoordinates.push(obj)
 				}
-				if (checkThatTheFieldIsFree(newCoordinates)) {
+				if (module.exports.checkThatTheFieldIsFree(newCoordinates, field)) {
```
Comment (zbstof) [inline tetris.js]: replace `module.exports.` with `this.`

File: tetris.js [hunk 40]
```diff
@@ -203,6 +175,34 @@ module.exports = {
...
+			}
+		}
+		return true;
+	},
+
+	allItemsOnTheField: function (arrayOfCoordinates, field) {
+		let lengthField_X = field[0].length;
+		let lengthField_Y = field.length;
+		for (let coordinate of arrayOfCoordinates) {
+			if (
```
Comment (zbstof) [inline tetris.js]: join this line and the next

File: tetris.js [hunk 41]
```diff
@@ -203,6 +175,34 @@ module.exports = {
...
+				(coordinate.x < 0 || coordinate.y < 0) ||
+				(coordinate.x >= lengthField_X) ||
+				(coordinate.y >= lengthField_Y)
```
Comment (zbstof) [inline tetris.js]: join this line and the next
this is confusing to leave if statement round brackets by its own

File: tetris.js [hunk 42]
```diff
@@ -430,5 +430,25 @@ module.exports = {
...
+			if (module.exports.checkThatTheFieldIsFree(tetrominoRotateResult, field)) {
+				this.rotationPhase = module.exports.incrementPhase(this.typeOfTetromino, this.rotationPhase)
+				this.coordinates = tetrominoRotateResult
+				return tetrominoRotateResult
+			} else {
+				return this.coordinates
+			}
+		}
+		moveTetromino(){
+
```
Comment (zbstof) [inline tetris.js]: and implemetation

File: tetris.js [hunk 43]
```diff
@@ -430,5 +430,25 @@ module.exports = {
...
 		}
+	},
+	Tetromino: class {
+		constructor(typeOfTetromino, rotationPhase, coordinates) {
+			this.typeOfTetromino = typeOfTetromino
+			this.rotationPhase = rotationPhase;
+			this.coordinates = coordinates;
+		}
+		rotateTetromino(field) {
+			let tetrominoRotateResult = module.exports.rotateTetromino(this.typeOfTetromino, this.rotationPhase, this.coordinates)
```
Comment (zbstof) [inline tetris.js]: this. rotateTetromino

File: tetris.js (hunk 39 above)
Comment (PashaBeshchuk) [reply tetris.js]: that doesn't work :(

File: tetris.js (hunk 40 above)
Comment (PashaBeshchuk) [reply tetris.js]: I don't understend  :(

File: tetris.js (hunk 43 above)
Comment (PashaBeshchuk) [reply tetris.js]: that doesn't work :(

File: tetris.js (hunk 40 above)
Comment (zbstof) [reply tetris.js]: Try
https://stackoverflow.com/a/51762551

File: test/test-moveTetromino.js [hunk 44]
```diff
@@ -738,7 +802,9 @@ describe("Move tetromino", function () {
         `)
     })
     it("O tetromino can not move, item too down", function () {
```
Comment (zbstof) [inline test/test-moveTetromino.js]: your test says that tetromino cannot move, but expected is different from coordinatesTetromino

Comment (zbstof) [reply test/test-moveTetromino.js]: Also, where is assertion?

File: tetris.js [hunk 45]
```diff
@@ -192,11 +192,7 @@ module.exports = {
 		let lengthField_X = field[0].length;
 		let lengthField_Y = field.length;
 		for (let coordinate of arrayOfCoordinates) {
-			if (
-				(coordinate.x < 0 || coordinate.y < 0) ||
-				(coordinate.x >= lengthField_X) ||
-				(coordinate.y >= lengthField_Y)
-			) {
+			if ((coordinate.x < 0 || coordinate.y < 0) || (coordinate.x >= lengthField_X) ||(coordinate.y >= lengthField_Y)) {
```
Comment (zbstof) [inline tetris.js]: format

File: tetris.js [hunk 46]
```diff
@@ -447,8 +443,13 @@ module.exports = {
 				return this.coordinates
 			}
 		}
-		moveTetromino(){
-			
+		moveTetromino(field, shift){
+			let shiftTetrominoResult = module.exports.shiftCoordinates(this.coordinates, shift)
+			if(module.exports.checkThatTheFieldIsFree(shiftTetrominoResult, field)){
+				return shiftTetrominoResult
```
Comment (zbstof) [inline tetris.js]: this is not the full algorithm

Add tests that move 2 times in different directions, like down-left, right-down, left-right and check the result is as expected

File: test/test-moveTetromino.js [hunk 47]
```diff
@@ -0,0 +1,829 @@
...
+        let phase = "up";
+        let coordinatesTetromino = tetris.getCoordinates(`
+            -XX--
+            -X---
+            -X---
+            -----
+        `)
+        let field = tetris.buildField(`
+            -----
+            --X--
```
Comment (zbstof) [inline test/test-moveTetromino.js]: this field is not in the way of moving up - test name does not reflect what you are checking

File: test/test-moveTetromino.js [hunk 48]
```diff
@@ -0,0 +1,829 @@
...
+        let expected = tetris.getCoordinates(`
+            -X---
+            -X---
+            XX---
+            -----
+        `)
+    })
+    it("J tetromino can not move, item too right", function () {
+        let typeOfTetromino = "J";
+        let phase = "down";
```
Comment (zbstof) [inline test/test-moveTetromino.js]: moving down should be possible in this case

File: test/test-moveTetromino.js [hunk 49]
```diff
@@ -0,0 +1,829 @@
+var assert = require('chai').assert;
```
Comment (zbstof) [inline test/test-moveTetromino.js]: You are missing assertions in all your tests

File: test/test-moveTetromino.js (hunk 48 above)
Comment (PashaBeshchuk) [reply test/test-moveTetromino.js]: but I move tetromino left

Comment (PashaBeshchuk) [reply test/test-moveTetromino.js]: but I move tetromino right

File: tetris.js [hunk 50]
```diff
@@ -430,5 +426,31 @@ module.exports = {
...
+			} else {
+				return this.coordinates
+			}
+		}
+		moveTetromino(field, shift){
+			let shiftTetrominoResult = module.exports.shiftCoordinates(this.coordinates, shift)
+			if(module.exports.checkThatTheFieldIsFree(shiftTetrominoResult, field)){
+				this.coordinates = shiftTetrominoResult
+				return shiftTetrominoResult
+			}else{
```
Comment (zbstof) [inline tetris.js]: format this file

**Pull request merged**